### DISCUSSION TITLE
Give us the power of code-first-i18n

### DISCUSSION THREAD

**Discussion by earloc** [earloc/TypealizR#78] [Ideas]
# introduction
Inspired by the (_at time of writing this_) still ongoing discussion over at #70 (and partially #38) I´m currently starting to think about *code-first-i18n* - similar to what [code-first-databases] are offering to us.
I´ll try to sketch out the idea emerging in my mind.

When it comes to translating apps, the first thing a dev needs to do is place some simple string (`"Hello App"`) or templated-string (`"Hello {0}"`)  somewhere within the app. 
> I´ll refer to those as `localizables`

In .net-land this probably would be utilizing plain old `resx`-files (or `resw` in UWP, etc.).

There are other options, indeed, like [OrchardCore.Localization] for [po-files]  (which I currently really need to explore more) or other offerings from - mostly paid - third parties (like [poeditor] or [localizely], just to name a few).

As I understand currently, all those offerings mostly are based on the fact that there IS already some sort of tech-stack native localization file (in .net-land some `resx`- or maybe an `xliff`-file), which is used as the point of interchange.

Getting to those files currently isn´t very intuitive or "developer-friendly", at least from my personal point-of-view.

So, what am I proposing?

# code-first i18n

From a developer´s point-of-view, all I really want is some sort if `Interface` I can inject and leverage existing IDE refactorings to introduce new `localizables`.

Why isn´t introducing a new `localizable` as easy as writing a new method (inlcuding parameters, if needed) and hitting `CTRL+.` and selecting `Generate method`:
![hello-code-first-i18n](https://user-images.githubusercontent.com/8486862/213010820-d5cde9bb-eaf6-4545-a0db-101559cdf3d4.gif)

## nice idea, but how does this help?
Exactly, this doesn´t really help now, as there´s nowhere some usable implementation around. 
This is where `TypealizR` could kick in. By marking the `Interface` with something like `[Typealize]`, a `source-generator` could emit a discrete implementation, as well as some sort of actual default exchange-file for storing actual `localizables`. This could be an ordinary `resx`-file OR a `xliff` / `po`, whatever. What the actual implementation really does, is currently not specified. Could be looking up from some file, or just honoring the requested `Culture` and deliver a static string - doesn´t really matter.

Inspired by what [MAT] does, this could then be used to generate even more files intended for interchange with translators or tools like [poeditor] / [localizely] / [whatever]. 
There´d need to be - of course - some mechanism to re-import such an exchange-file, to support workflows involving external tools / translation-providers. But this looks kind-of solvable, atm. Even if that would involve taking `translatables` back and incorporating those into the static strings, which might be returned by generated code.

Any opinions?

[code-first-databases]:https://www.google.de/search?q=code+first+database
[OrchardCore.Localization]:https://www.nuget.org/packages/OrchardCore.Localization/
[po-files]:https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html
[poeditor]:https://poeditor.com/
[localizely]:https://localizely.com/
[whatever]:https://www.google.de/search?q=i18n+tool
[MAT]:https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-mat

**Comment by earloc** [comment]
Here´s a possible "workflow", at least how I imagine code-first-i18n to work, atm.
> the yellow spots are just for reference, where other toolings could pick up and be incorporated in the overall flow.

![CodeFirst drawio](https://user-images.githubusercontent.com/8486862/213250399-92d672ba-caae-4fca-96de-fe37e908d9a2.png)

1. A develper authors a plain-old `Interface` with the editor of choice. (This might need to be marked with an appropriate `Attribute`)
```csharp
[CodeFirstLocalizable]
interface ILocalizables
{
    ///<summary>
    /// Hello '<paramref name="world"/>'
    ///</summary>
    /// <param name="world">the name of the world to be greeted</param>
    /// <returns></returns>
    LocalizedString Hello(string world);
}
```

2. A `source-generator` analyzes said interface
3. The `source-generator` generates a class which implements the `Interface`. It may use the xml-comment to populate a meaningful default translation, otherwise a generated string could be deduced from the member´s signature (`properties` or `methods`). Implementation details do not matter, atm. Could be generating a hard-coded string, or falling back to e.g. `IStringLocalizer<T>`...dunno, yet.
```csharp
partial class Localizables : ILocalizables
{
    public LocalizedString Hello(string world) => ...;
}
```
4. A yet-to-be developed CLI kicks in and reads the generated classes. Either via out-of-process reflection or roslyn. 
5. The CLI synchronizes all strings into plain-old `Resx`-files (Other file-formats are definetly thinkable)., possibly placing some breadcrumbs along the way to ease up sync on subsequent invocations. We now have a default file that can be modified as needed.
6. Tools like [MAT] or whatever could enhance the overall process by doing their thing(s). For [MAT], this is to generate an `XLIFF`-file out of all `Resx`-files within a project and assembling those into locale specific `XLIFF`-files.
7. Those files are then used to populate locale specific `Resx`-files
8. [MAT] further synchronizes things into existing `XLIF`-files, which then can get translated by the developer or a third party (tool).
9. The changes get reflected within the locale specific `Resx`-files

[MAT]:https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-mat

**Comment by MarkPflug** [comment]
I'm still trying to wrap my head around your proposal here, so I don't have much to comment about it yet. However, I thought I'd point out a similar effort that I've worked on in the past: https://github.com/MarkPflug/Sylvan.BuildTools.Resources. It is a package that provides localization support very similar to what .resx files offer, but rather than using XML, it uses JSON (.resj) files. The project readme outlines a few of the advantages over resx, if you're interested.

My package uses build-time code generation (an MSBuild task), rather than compile-time code gen (source generator) as your project uses. I don't see a huge advantage to one over the other. The one limitation I found with source generators was that it isn't ([yet](https://github.com/dotnet/roslyn/issues/49935)) possible to add embedded resources to the compilation. My project only came about as a proof-of-concept, for a larger project I was working on professionally, so I haven't really been adding much to it recently.

Resx uses a combination of design-time (Visual Studio custom tool) for code generation and build-time (MSBuild task) for creating the embedded resources. I was trying to find a way to do something similar that didn't require Visual Studio, so other developers could use VSCode and still be able to manage the resj files. 

One idea from your project that I really like is the strongly-typed formatter method design. My project, much like RESX, only produces localized strings, so if a localized string is a format string, it is up to the consumer to apply the correct formatting, which wouldn't have any compile time checks. `string.Format(MyResources.GreetingFormat, username)`, for example. That code doesn't have any checks that the correct number/types of arguments is supplied for the format string. It would be a nice enhancement to also generate a formatter method on the resource accessor class so I could also write: `MyResources.Greeting(username)`, which could provide compile-time checks that the arguments were correct.

It looks like your project is mostly targeted at the new `IStringLocalizer`, which I'm only passingly familiar with from ASP.NET examples I've seen. My understanding is that it is meant to provide an abstraction on top of a concrete implementation such as .RESX. This makes me wonder if your projects might provide a layer on top of what I've created. 

Anyway, just wanted to say hi. I'll be keeping an eye on your project's progress.

Cheers,
-Mark

**Reply by earloc** [reply]
Hey @MarkPflug, nice to have you here.
Let me say that using your [Sylvan](https://github.com/MarkPflug/Sylvan)-stuff recently really was a breeze to work with. ;)

I happen to just have a look at your `ResJ`-approach a couple of hours ago. I really like the idea of having a less verbose file-format for providing `localizables`.

When it comes to localizing an app, the supported features always felt a bit as a 2nd- or 3rd-class citizen in .net-land. This is why I came up with `TypealizR` in the first place (I recently had to swoop over a Blazor-App and refactor the used `localizables` there, which was kind of a PITA. So I thought about what might really add some benefit there). 
This discussion here is just some natural evolvement which happens when tackling such things with inputs from different viewpoints, like yours. Greatly appreciated!

`IStringLocalizer` has some overall advantages over the classic `*ResXFileCodeGenerator`-approach, which is present since the very early days of .NET. 
See this [SO-article](https://stackoverflow.com/questions/69129051/resx-in-blazor-wasm-what-is-the-issue-with-the-old-static-way-of-using-the-re/74272357#74272357) here.
I recently played with adding support for ordinary `*ResXFileCodeGenerator`-enabled `Resx`-files, but presumably ditched the idea as even MS seems to only advocate for using `IStringLocalizer` these days. And, well, there were some stupid technical reasons for why this wouldn´t work. See #23.

However, introducing other file-formats as an input for generating sources has been *the* single requested feature since TypealizR got some traction. Sadly, I personally didn´t found some eligible approach adopting those...See #38 (`*.po*`-files)  and #70 (`*.xliff`-files). But I certainly like the idea to extend upon other file formats.

One Idea for `ResJ` would be to provide a custom `IStringFormatter` for such kind of files, as those seem to be 1:1 compatible to classic `ResX`-files (in terms of features). Then, it should be relatively easy for TypealizR to use those files instead of `ResX` to generate extensions-methods. If you´re interested, plz have a look at what the [OrchardCore.Locaization]-peoples have done so far making it feasible to use non-ResX in "modern" apps. TypealizR currently really only is a generator on top of an existing, well understood mechanism to provide `localizables` in a type-safe way. This might change with `code-first-i18n`, which I´m personally still trying to wrap my head around, also, like yourself ;). But the topic - as a whole - feels very intriguing to me, which will surely keep me busy some more time ;) stay tuned.

[OrchardCore.Locaization]:https://github.com/CMSCollection/OrchardCore/blob/9db4098855de610d95834b18eb3adb205cf52cd4/src/OrchardCore/OrchardCore.Localization.Core/ServiceCollectionExtensions.cs

**Reply by earloc** [reply]
Oh, and btw, I think favoring `source-generators` over `msbuild-tasks` might have the advantage of a much faster user-feedback:
![TypealizR](https://github.com/earloc/TypealizR/blob/main/docs/assets/demo_typealize_translation_initial.gif?raw=true)

I don´t think such responsiveness could be achieved with `msbuild`, as there would be a complete `rebuild´-cycle needed in order to get the generated stuff ready in place. And this is real-time up there, no cheating ;)

**Reply by MarkPflug** [reply]
You would be right, if not for one super-secret trick. It allows the resource strings to show up in intellisense as soon as the resj file is saved. It essentially runs a subset of the full build upon file save. I've written about it here: https://markpflug.github.io/2018/08/23/JsonResource.html

If it required a full-rebuild to get the resources to show up in Intellisense, then I would consider it un-usuable.

If you give it a try, be aware that Visual Studio might (unhelpfully) add an element to your .csproj that might prevent it from working:

```xml
 <ItemGroup>
    <JsonResource Remove="Resources.resj" />
  </ItemGroup>
```

If that shows up in your .csproj, just delete it, and save. Should fix the problem. I have an open issue on the repo: https://github.com/MarkPflug/Sylvan.BuildTools.Resources/issues/9, but haven't yet figured out a way to prevent VS from inserting it.

**Reply by earloc** [reply]
N1, I'll definitely dive into that article, thx! ✌️

**Reply by earloc** [reply]
Quite a journey you had up there at https://markpflug.github.io/2018/08/23/JsonResource.html, thx for sharing this to us!

I thought about a possibility to extend over existing `*ResXFileCodeGenerator`-enabled files again, which also could be applied to things generated by [Sylvan.BuildTools.Resources], as both generators produce static accessors.

Would be cool if you have a look over at https://github.com/earloc/TypealizR/issues/23#issuecomment-1396604393, if such an approach would be feasable for you. (Ignore the fact that it´s based on `Resx`, I guess applying this to `Resj` is probably not that hard.






[Sylvan.BuildTools.Resources]:https://github.com/MarkPflug/Sylvan.BuildTools.Resources

**Reply by earloc** [reply]
@MarkPflug 

I just had a quick look into the generated code from [Sylvan.BuildTools.Resources](https://github.com/MarkPflug/Sylvan.BuildTools.Resources) .
`TypealizR` currently expects parameters (and type-annotations) within the resource´s `key`.

This does not play well with [Sylvan.BuildTools.Resources](https://github.com/MarkPflug/Sylvan.BuildTools.Resources)

```json
{
  "Strings": {
    "Greeting": "Hello, Resj",
    "Hello {world:s}": "Hello {0}"
  },
}
```

produces this method signature:
```csharp
    public static string Greeting
    {
        get
        {
            return ResourceManager.GetString("Greeting", resourceCulture);
        }
    }
    public static string Hello {world:s}
    {
        get
        {
            return ResourceManager.GetString("Hello {world:s}", resourceCulture);
        }
    }
```

Any idea how to sort this out?
I could help out and maybe provide `TypealizRs` sanitizing mechanism in a package like [TypealizR.Core] or something. Or you could just play copy-cat at this early stage.

With this in place, we where very close to combine our approaches. A `resx`-based POC is in the makes, currently.

**Reply by MarkPflug** [reply]
Thanks for the note. Indeed, the current expectation is that the key must be a valid C# identifier, because it is used directly to generate the accessor properties. Kinda makes me think that is will end up being incompatible with what your library expects, not sure.

https://github.com/MarkPflug/Sylvan.BuildTools.Resources/issues/15

**Reply by earloc** [reply]
Just commented on that issue, there. If you're open to adopt the needed changes, I might jump in and try to submit  a PR.

If that doesn't fit your overall design, we could also think about alternatives for augmenting those keys wit parameters and type-annotations to help [TypealizR] fiddling out what's going on 😉.

**Reply by MarkPflug** [reply]
Don't spend time on a PR right now. I want to give a bit more thought to this before committing to something. Busy with other stuff at the moment though, so it might be a while before I can give it the attention it deserves.

**Reply by earloc** [reply]
Hey @MarkPflug .

Checkout #88, I think with this in place we're not far away to also incorporate your `resj` format. Just need to think about how: hardcoded or via some kind of plugin-model , maybe.

**Reply by earloc** [reply]
Tracking work in #95

**Reply by earloc** [reply]
@MarkPflug maybe this helps with head-wrapping / grasping the idea behind `code-fist`-resources ;)
https://github.com/earloc/TypealizR/blob/main/docs/code-first.md

**Reply by MarkPflug** [reply]
Thanks for the reminder. I need to find some time to dig into this.

**Reply by earloc** [reply]
Just dropping in this here, so I'll later find it for reference:

https://medium.com/younited-tech-blog/json-based-blazor-localization-395b1b31e8ae

**Comment by ivanhofer** [comment]
Hi, I'm Ivan and I have done something similar to `TypealizR` in the `JavaScript` ecosystem: https://github.com/ivanhofer/typesafe-i18n

i18n is a pain for developers. Making it as effortless as possible is certainly a goal tooling should strive for. In my opinion the main problem is that the place where you call the translation function is not the place where your strings are stored (key-based translations). If you need to add a new string to the application, you need to come up with a new key, write the string into a dictionary-file and then call the translation function inside the code.

If I understand everything correctly, you want to eliminate the step of requiring a developer to manually write something into the dictionary file. He does not need to leave the file where he wants to call the translation function. I like that approach as it eliminates a pain-point when dealing with i18n.

I'm currently working on a draft of a concept to make i18n more accessible to web developers: https://github.com/pipeli18ne/RFC
The goal is that you just annotate code parts that you want to be internationalized, and tooling does the rest for you. No need to manually handle dictionary files, no need to configure it for a certain framework etc.. Everything could be automated to make it as effortless as possible to use.

**Reply by earloc** [reply]
Hey @ivanhofer.

Had a look on your package there a while ago. This really looks like a sustainable way for handling this in typescript, at least when you have choices and are not bound to some framework stuff (like Angular or similar) 

Yeah, that really is the key takeaway which I'm trying to achieve with this proposal here - reducing the occasions a dev needs to leave their natural habitat (which is, the code, of course 😉) 

I already have a POC for some of the heavylifting needed to be done in progress. Still curios, if the approach will hold position when facing real-world problems, though 🤗.

And I'll definetly check out your link.

**Reply by earloc** [reply]
Funny, though, how be both came up with the same style for specifying parameters `Hello {world:string} ` in translations, independently from each other 😝✌️

**Reply by ivanhofer** [reply]
Yeah 😅 At least for the `TypeScript` world, this felt quite natural :)

**Reply by earloc** [reply]
@ivanhofer if you need to stick your head up from looking at typescript localizations and peek into .net-land, this might better explain what I´m trying to achieve:
https://github.com/earloc/TypealizR/blob/main/docs/code-first.md
Cheers

**Reply by ivanhofer** [reply]
I'll take a look at this when I find some time :)

**Reply by ivanhofer** [reply]
I did read it. I have to admit that I don't know anything about `.NET` (even if it was my first language/platform when I started programming😅).

Overall it looks good to me. The instructions seems easy to follow. Where I had troubles is where you mention `structured xml comments`. In the example above that point, you mark that `codeFirst.Hello` outputs "Hello". Is this auto generated from the interface? For me it is not really clear where this comes from.

And where you show the xml comments, why does `Greet` have a `<returns></returns>` at the end but `Hello` not?
You are also using `paramref` and `param` there. Is `param` needed or could it also work if you just define `paramref`? 

I also don't know anything about `*.resx` files, but why are the parameters `0` and `1` and not `userName` and `planetName` like in the interface?

**Reply by earloc** [reply]
Thx @ivanhofer, I really appreciate your time and effort on this. Thank you so much.

I revised the documenation a bit to clarify / simplify things over at #105.
>In the example above that point, you mark that codeFirst.Hello outputs "Hello". Is this auto generated from the interface?

Yes, exaclty. That´s one of the main points where the generator kicks in and let you just start with a naked, un-customized `localizable`, to not break dev-flow.

>And where you show the xml comments, why does Greet have a <returns></returns> at the end but Hello not?
> You are also using paramref and param there. Is param needed or could it also work if you just define paramref?

Those are just left overs from what you get within Visual Studio once you type `///`. The sample will also work exactly the same without those `paramref` and `return`-elements.
See [xmlcomments](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/), which is an editor-assisted way to specify code-comments since `.nets`-initial incarnation back in 2002, I guess. Very similar to what e.g. [jsdoc](https://jsdoc.app/about-getting-started.html) do.

>I also don't know anything about *.resx files, but why are the parameters 0 and 1 and not userName and planetName like in the interface?

This is because, internally, the implementation ultimately utilizes a call to [string.Format](https://learn.microsoft.com/en-us/dotnet/api/system.string.format), which happens to have no support for named paramaters, only positional ones. That´s just how it works, sadly ;( . But I´m thinking of a way to let users customize even this, similar to what the `resource-first`- approach currently offers, see #39.

**Reply by ivanhofer** [reply]
> I revised the documenation a bit to clarify / simplify things over at #105.

Great!

> and let you just start with a naked, un-customized `localizable`, to not break dev-flow.

Nice detail!

> The sample will also work exactly the same without those `paramref` and `return`-elements.

Maybe you could mention this somewhere.

> This is because, internally, the implementation ultimately utilizes a call to [string.Format](https://learn.microsoft.com/en-us/dotnet/api/system.string.format), which happens to have no support for named paramaters, only positional ones

This may lead to problems when you have multiple parameters in a message and the position of those change in a translation. Named parameters solve this issue, but I guess you already know that.

**Reply by earloc** [reply]
Gnaah, I really meant to say `index-based`, rather than `positional` 😉.  And that's only true for the values itself. So it's absolutely ok, if the order between different languages might change during the translations Process. The index gets fixed based on the positional order arguments get passend down to that internal call. And this static code won't change during translation at all - and when, then only on purpose. 😇

But yeah, having just those indexes at hand really kind of sucks  a little bit 😉.

**Comment by earloc** [comment]
Hey ya´ll.

I recently made some huge steps making this happen, first MVP is about to land soon. See #88.

**Comment by earloc** [comment]
https://github.com/earloc/TypealizR/releases/tag/v0.9.2 just landed.

See also https://www.reddit.com/r/dotnet/comments/10ufygg/introducing_codefirsti18n_internationalization/?utm_source=share&utm_medium=web2x&context=3

**Comment by samuelstroschein** [comment]
> Inspired by what [MAT](https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-mat) does, this could then be used to generate even more files intended for interchange with translators or tools like [poeditor](https://poeditor.com/) / [localizely](https://localizely.com/) / [whatever](https://www.google.de/search?q=i18n+tool).
There´d need to be - of course - some mechanism to re-import such an exchange-file, to support workflows involving external tools / translation-providers. 

I introduce https://github.com/inlang/inlang. One of the apps lets translators manage translations directly on a git repo without the need for re-import/sync.  

There seems to be a general overlap in the goal of TypealizR and inlang: make i18n easier, especially for developers.

**Reply by earloc** [reply]
Hey, cool stuff there. Will definetly check out how this might suite in the overall vision. Thx for spreading the word! ✌️

**Comment by francisphn** [comment]
About two years ago, I read [an interesting blog post by the folks at Canva on their internationalization engineering practice](https://www.canva.dev/blog/engineering/how-to-design-in-every-language-at-once/). They use the code-first process as you've described here. For Canva which is a big platform that needs to support 100+ languages, code-first-i18n has appeared to emerge as the most scalable way to implement i118n.
